feat: player waiting room modal and dialog component added

ashleybhandari opened the pull request:
- Created general modal component, added to storybook
- Started Waiting Room
  - Can edit team name
  - Responsive and can't be closed

![Screenshot 2025-03-12 123100](https://github.com/user-attachments/assets/40131a53-ee9e-47da-9423-90c8e4fb5e51)
![image](https://github.com/user-attachments/assets/906256f0-a384-4f88-b679-56165f676920)

ashleybhandari committed:
Install shadcn dialog #11

ashleybhandari committed:
Create dialog story #11

ashleybhandari committed:
Create modal
- Options to default to open and to disable closing
- Header can be hidden
#11

ashleybhandari committed:
Create skeleton for waiting room
- Static setup for establishing component hierarchy (excl. QR code)
#11

ashleybhandari committed:
Removed ModalHeader #11

Yonava commented:
Can we go from this skeleton ui to something more closely resembling the final state of the ui by Monday, especially considering how long it has been since this project has been going. A couple comments I have is firstly

* name your props types something other than just Props, instead opting for a fuller name like WaitingRoomProps,

* organizing your parent component into sub component structures like WaitingRoomPlayer and WaitingRoomTeam especially when the type of player or any of the implementation details of the component have not been scaffolded yet may be some premature optimization you want to avoid doing

* premature optimization  also applies to styles like `md:flex-row` when the basic UI hasn't even begun taking shape.

The workflow should look something like making a parent, adding some stuff to it until something seems like it could be broken off into its own thing, making a component for that thing, and then once everything is looking pretty close to done, adding some breakpoints to make it work on different screen sizes.

ashleybhandari committed:
Rename Props type to be more specific to the component it's being used with #11

ashleybhandari committed:
If Modal has disabled closing, it won't close when the escape key is clicked #11

ashleybhandari committed:
Create EditableInput component (text with a button that toggles editing) #11

ashleybhandari committed:
Add avatars and "remove" buttons to waiting room players
- Created User type
- Created user mock data
- Created ProfilePicture component (using shadcn Avatar)
[#11](https://github.com/Yonava/code-swap/issues/11)

ashleybhandari committed:
Modal grows wider with content

ashleybhandari committed:
EditableInput edit button disappears while editing
- Makes editing look cleaner
- Empty string is replaced with last valid value
[#11](https://github.com/Yonava/code-swap/issues/11)

ashleybhandari committed:
Modal always has a dialog description, regardless of whether description has content (resolves console error) [#11](https://github.com/Yonava/code-swap/issues/11)

ashleybhandari committed:
TAB saves changes and ESC discards them to align better with expected behavior (EditableInput)
#11

ashleybhandari committed:
Rename User as Player to keep verbiage consistent #11

ashleybhandari committed:
Indicate empty slot(s) in the waiting room when a team isn't full #11

ashleybhandari committed:
Replace X buttons with a Start Game button #11

ashleybhandari committed:
Merge branch 'main' into dialog

Yonava commented on client/src/components/EditableInput.tsx:
```diff
@@ -0,0 +1,65 @@
+
+  return (
+    <>
+      {isEditing && (
```
We can combine the expressions starting on line 42 and 52 using a ternary

Yonava commented on client/src/components/Modal.tsx:
We are calling the file and the component modal while implementing a dialog. [Heres an article about the difference](https://dev.to/iam_timsmith/dialogs-vs-modals-is-there-a-difference-210k)

Because we are using Shadcns Dialog I favor just naming everything to reflect that

ashleybhandari replied:
Yeah that makes sense, I was just scared of interfering with shadcn's naming.

Yonava commented on client/src/components/Modal.tsx:
```diff
@@ -0,0 +1,70 @@
+  DialogTrigger,
+} from '@/components/ui/dialog';
+
+type ModalProps = {
```
can we add some tsdocs to these props. Some of them could use a little explainer

Yonava commented on client/src/components/Modal.tsx:
```diff
@@ -0,0 +1,70 @@
+    }
+  };
+
+  const header = (
```
If this returns jsx, it should probably be capitalized by convention

Yonava replied:
and also calling it something like HeaderContent may a slight improvement to readability

Yonava commented on client/src/components/Modal.tsx:
```diff
@@ -0,0 +1,70 @@
+        onEscapeKeyDown={handleClose}
+        className={cn('!max-w-fit', { '[&>button]:hidden': disableClose })}
+      >
+        {!hideHeader && header}
```
ternary

Yonava commented on client/src/components/ProfilePicture.tsx:
```diff
@@ -0,0 +1,15 @@
+import { Avatar, AvatarFallback, AvatarImage } from '@/components/ui/avatar';
+
+type ProfilePictureProps = {
+  src?: string;
```
Should `src` be optional?

ashleybhandari replied:
People without an account may not have a profile picture, so I made it optional. The avatar falls back to a little "</>" in a circle in this case.

Yonava replied:
If we dont have a profile picture, I dont think we should be rendering this component. What may be better is to create a "empty slot" componnet or something then have a default picture which we feed into here

ashleybhandari committed:
Renamed EditableInput as EditableText because previous name didn't make sense #11

Yonava commented on client/src/components/waiting-room/WaitingRoom.tsx:
```diff
@@ -0,0 +1,30 @@
+      isOpen
+      disableClose
+      hideHeader
+      className="flex flex-col gap-10 w-[80vw] sm:w-96 md:w-2xl"
```
for dialogs/modals like this, its typically standard to have width be something like min(500px, 80vw)

Yonava replied:
that way the modal has a hard cap on how big it can get while taking up a reasonable amount of space on all screen sizes below the larger variants

Yonava commented on client/src/components/waiting-room/WaitingRoom.tsx:
```diff
@@ -0,0 +1,30 @@
+      hideHeader
+      className="flex flex-col gap-10 w-[80vw] sm:w-96 md:w-2xl"
+    >
+      <div className="flex flex-col md:flex-row gap-y-12">
```
flex direction is row by default. Just put a breakpoint prefix to flex-col so it is only active when needed and get rid of md:flex-row

ashleybhandari replied:
Tailwind follows a mobile-first breakpoint system. I only want flex-col when the screen is small, so flex-row must be specified for md and above.
"What this means is that unprefixed utilities (like uppercase) take effect on all screen sizes, while prefixed utilities (like md:uppercase) only take effect at the specified breakpoint and above."
https://v2.tailwindcss.com/docs/responsive-design#mobile-first

Yonava replied:
good to know!

Yonava commented on client/src/components/waiting-room/WaitingRoomTeam.tsx:
```diff
@@ -0,0 +1,44 @@
+export const WaitingRoomTeam = ({ defaultName }: WaitingRoomTeamProps) => {
+  const [players, setPlayers] = useState([MOCK_PLAYERS[0], MOCK_PLAYERS[1]]);
+  const emptySlots = [];
+  const slotClasses = 'h-10 flex items-center';
```
why are these classes not inlined?

ashleybhandari replied:
Full and empty slots should take up the same height. I figured they'd have more styling in common later on so I just made a variable and applied it to both.

Yonava replied:
this may be a good use case for creating a wrapper component called PlayerSlot, applying the styles there and creating two variants, one for EmptyPlayerSlot and FilledPlayerSlot

ashleybhandari replied:
oo good idea

Yonava commented on client/src/components/waiting-room/WaitingRoomTeam.tsx:
```diff
@@ -0,0 +1,44 @@
+  defaultName: string;
+};
+
+export const WaitingRoomTeam = ({ defaultName }: WaitingRoomTeamProps) => {
```
In this component there is a lot of fanciness going on with how empty slots are computed including with loops, maps etc.

for this case specifically, I would advise against adding this logic in favor of keeping things dumb. Every team can only have two players, and only two players due to the core swapping mechanic. Because the constant 2 is so small and the number is such a constant in the design of this application, I would get rid of the loops and logic all together in favor of just a "firstPlayer" "secondPlayer" state. if the player seat hasnt been filled, it can be set to undefined, if its undefined we can render the Empty slot

ashleybhandari committed:
Rename Modal to include "Dialog" #11

ashleybhandari committed:
Replace players state with player1 and player2 states to simplify code #11

ashleybhandari committed:
Add TSDoc to DialogBox to clarify props #11

ashleybhandari committed:
Change Waiting Room dialog sizing to be more standard #11

ashleybhandari committed:
Profile picture src is required, implementers have to supply a default #11

ashleybhandari committed:
Storybook stories
